sds_in_a_box/SDSCode/api.py

```diff
@@ -0,0 +1,76 @@
+import json
+import os
+import requests
+import time
+from jose import jwk, jwt
```

Comment:
Why did you choose to use jose instead of boto3 to take care of the authentication for you?

Reply:
I actually couldn't figure out how to have boto3 do it! If you do this on another project let me know.  But on this page: https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-verifying-a-jwt.html it basically tells you exactly what you need to do to verify. 

There are a bunch of AWS SDK operations here: https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_Operations.html but I couldn't see one that verified a token.  There is one at the bottom called "VerifySoftwareToken" but that seems to involve MFA one time codes.

Reply:
The web team does it on their end and uses the AWS specific frameworks on their end.

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cognito-idp.html#CognitoIdentityProvider.Client.initiate_auth

https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_InitiateAuth.html

Or actually, are you trying to add Cognito as an auth mechanism to all APIs? We can let AWS handle that by adding an authorizer in CDK.

Brian M set that up in TREC for us and had it working, but we got rid of Cognito because people didn't want to deal with a login:
https://github.com/SWxTREC/swxtrec-cdk/commit/7edb962a89868a10a545a62df30ada2b53faeec9

```python
        # Add Cognito auth requirement to API gateway
        authorizer = apigw2_auth.HttpUserPoolAuthorizer(
            "APIGWAuthorizer",
            authorizer_name="APIGWAuthorizer",
            pool=user_pool,
        )
```

Reply:
Yeah we're adding it as an authorization mechanism for all the APIs.  The current plan is that if they are a team user, they can get access to non-public data.  If they don't add an authorization token in their header, then they can still access the public data.  

As of right now we aren't using API Gateway, we're invoking the lambda functions directly via HTTPS calls, so the API Gateway authorizer won't work with it.  The advantage of invoking the lambda directly is 1) Less infrastructure in general to worry about and 2) The limits on API gateway don't apply to invoking the lambda directly (for example, the 29 second timeout with API gateway)

Reply:
OK, sounds good for now! I don't know anything about the Lambda URLs yet and asked about them on the AWS working group call yesterday, but no one there had experimented with them either. So it sounds like you're breaking new ground and will be the expert here :)

> 1) Less infrastructure in general to worry about 

I actually think you get a lot from APIGW for as simple as it is to implement and it can handle DDOS attack vectors and other security things a little easier.

> 2) The limits on API gateway don't apply to invoking the lambda directly (for example, the 29 second timeout with API gateway)

I think this is probably the important reason why you'd want this to be urls rather than APIGW. Although I'd also say that as a user I get annoyed if my request is taking longer than 30 seconds and not sending me any streaming data back or things to let me know it is alive still, so if we're planning on long running Lambdas we should think if there is a better way to handle that situation. (I thought you did something like this on EMM where you'd put the data into a bucket for the user to download and send them a link to that instead of sending it directly back to the user)